### DISCUSSION TITLE
Two-Step Verification on Garmin Accounts

### DISCUSSION THREAD

**Discussion by iggbom** [philosowaffle/peloton-to-garmin#408] [Q&A] [answered]
Anyone know if there's a way to handle that?

**Comment by philosowaffle** [comment]
*02/04/2023 Edit: See #410 for progress updates*
*02/12/2023 Edit: See marked answer for solution*

*original:*
I honestly didn't even know Peloton offered 2fa. Is it done via an Email/Texted TOTP or can you use any app (like Authenticator) for generating the code?

**Reply by iggbom** [reply]
Don't think Peloton does offer MFA/2FA but I was asking about Garmin. They do using email or SMS code as the 2nd factor.

**Reply by philosowaffle** [reply]
I totally did not read/process the title of this thread, that's what I get for commenting before coffee :).  I can make a dummy Garmin account, enable 2fa, and play around with it to see what the options are (if any).

**Reply by cbordman** [reply]
Garmin is rolling out ECG on some watches.  It requires 2fa to be enabled.

**Reply by iggbom** [reply]
The Garmin Jr. App requires 2FA also. So any user of that app needs to disable/enable 2FA to use P2G/Garmin Jr.

**Reply by philosowaffle** [reply]
I've got an open task to look into this more: https://github.com/philosowaffle/peloton-to-garmin/issues/410

**Reply by johnnybirdman** [reply]
And just found out the hard way that once you opt into the 2FA, you can’t undo it.. no way out.

**Reply by iggbom** [reply]
@johnnybirdman I have been switching back and forth so it's been possible for me. Here's what I've done

Log in here: https://www.garmin.com/en-US/account/
Then go to "Update Sign-On Settings"
Flip both SMS and E-mail to Off

**Reply by johnnybirdman** [reply]
@iggbom, thank you for the info.  I gave it a try but unfortunately it says that it’s “permanent”… unless you know how to get around that too?  
![FB26406A-A529-44C5-9331-941B02095959](https://user-images.githubusercontent.com/124098236/215894576-f3051fb3-57d6-4bf3-84e8-daf8d4029704.jpeg)

**Reply by iggbom** [reply]
@johnnybirdman I don't but looks like maybe it's related to what device you have associated with your Garmin account? You got an ECG enabled device?

https://forums.garmin.com/apps-software/mobile-apps-web/f/garmin-connect-web/321912/permanent-two-step-verification

**Reply by johnnybirdman** [reply]
Yes, I have ECG working on the Venu 2 Plus.  I tried to uninstall the ECG app as well, but that didn’t seem to change anything.

**Reply by jsnrthmp** [reply]
I just spoke with Garmin support. The permanent 2FA is due to the ECG app. As they release it for more watches, this will become more prevalent. They will not/are not able to reverse the setting, even if you remove the app.

**Reply by jsnrthmp** [reply]
The authentication can be completed via email

**Comment by philosowaffle** [comment, marked as answer]
1month and 1600 modified lines of code later.... and I think I have this working good enough to hand off to y'all 🎉

If you find a problem please reach out on this thread.

**Notes on Two-Step:**
1. You will not be able to use the GitHub Action option anymore, you'll have to switch to one of the other [install options](https://philosowaffle.github.io/peloton-to-garmin/install/)
2. If you run the Docker Console image, it needs to be run in Interactive `-i` mode ([example command](https://philosowaffle.github.io/peloton-to-garmin/install/docker-headless.html))
3. You will not be able to enable automatic/background syncing. Syncing will always be a manual process since you will have to provide your Garmin one time code

**Config changes:**
1. If you use `configuration.local.json` then there is a [new flag](https://github.com/philosowaffle/peloton-to-garmin/blob/master/configuration.example.json#L30) that needs to be set
2. If you use the WebUI, you can find the new flag under the Garmin section in Settings

**Console/Command Line Experience:**
1. When the TwoStep Verification setting is enable, P2G will
    1. Start the auth flow with Garmin
    2. You will receive an email or text from Garmin with your 2fa code
    4. You will be prompted by P2G to enter your code
1. Assuming auth is successful, P2G will continue as normal from here

**WebUI Experience:**
1. Whenever you click a "Sync" button, P2G will check if we need to go through the 2FA flow
2. If so, you will receive an email or text from Garmin with your 2fa code
3. P2G will prompt you with a modal to enter your 2fa code
5. Assuming auth is successful, P2G will continue as normal from here
 
**How to get these changes:**
 Consider yourselves beta testers before I officially cut a new release of the code.  With that in mind you can grab the version you need from the below sources:

1. [Windows exe](https://github.com/philosowaffle/peloton-to-garmin/suites/10931493086/artifacts/552845814)
6. Docker - Make sure you are pulling the `console-latest`/`api-latest`/`webui-latest` tag instead of `stable` or one of the release versions
7. Source - Make sure you've pulled down the latest from the `master` branch

**Reply by cbordman** [reply]
Hi, i just pulled with 'docker pull philosowaffle/peloton-to-garmin:latest' and it says the image is 7 weeks old.

**Reply by philosowaffle** [reply]
Ugh 🤦 , it looks like while doing some unrelated changes back in December (#358) I accidentally renamed the tag to `console-latest`. Try switching to that.

**Reply by cbordman** [reply]
Here is the output i get after starting the container.  I am receiving the code via text but the container isn't stopping at the prompt to allow me to enter it. 


> [12:40:55 INF] *********************************************
[12:40:55 INF] P2G Version: p2g_console 3.4.0
[12:40:55 INF] Operating System: Unix
[12:40:55 INF] OS Version: Unix 5.15.49.0
[12:40:55 INF] DotNet Runtime: Unix 5.15.49.0
[12:40:55 INF] Docker Deployment: True
[12:40:55 INF] *********************************************
[12:40:55 INF] Application started. Press Ctrl+C to shut down.
[12:40:55 INF] Hosting environment: Production
[12:40:55 INF] Content root path: /app
[12:40:56 INF] *********************************************
[12:40:56 INF] *********************************************
Detected Garmin Two Factor Enabled. Please check your email or phone for the Security Passcode sent by Garmin.
Enter Code: Enter Code: Enter Code: Enter Code: Enter Code: [12:40:59 FTL] Uncaught Exception
Garmin.Auth.GarminAuthenticationError: Auth appeared successful but failed to find Garmin auth cookie.
   at Garmin.Auth.GarminAuthenticationService.CompleteGarminAuthenticationAsync(String loginResult, GarminApiAuthentication auth) in /build/src/Garmin/Auth/GarminAuthenticationService.cs:line 156
   at Garmin.Auth.GarminAuthenticationService.CompleteMFAAuthAsync(String mfaCode) in /build/src/Garmin/Auth/GarminAuthenticationService.cs:line 215
   at PelotonToGarminConsole.Startup.RunAsync(CancellationToken cancelToken)

**Reply by philosowaffle** [reply]
~Yeah I don't think you're gonna be able to run Console mode through Docker since it won't pause correctly to accept your input. If you want console mode you'll have to build from source or run the windows exe app if that's an option for you.  Or make the switch over to the Docker WebUI version.~

**Reply by philosowaffle** [reply]
@cbordman I figured out what you need to do for Docker Console:

```
docker run -i -v /full/path/to/configuration.local.json:/app/configuration.local.json philosowaffle/peloton-to-garmin:console-latest
```

You have to specify `-i` which is "interactive" mode.

**Reply by cbordman** [reply]
@philosowaffle that worked.  Great job !

**Reply by johnnybirdman** [reply]
Works great.  thank you!

**Reply by qpop01** [reply]
would there be any way to make this work via github actions? e.g. with a token or cookie with logged-in state? Github sites to produce a web interface for 2fa?

**Reply by philosowaffle** [reply]
Instead of investing time to find an even hackier way of using GitHub Actions / Sites, I would rather focus on solutions that will work without me having to write and support a whole bunch of new code and that will be more future-proof when changes like this happen.

A solution more along the lines of adding a one-click "Deploy to Heroku" or "Deploy to DigitalOcean" button (or whatever the popular hosting service is these days) would be ideal.

**Reply by qpop01** [reply]
Ah that makes sense. I'm mostly just sad about it because I'm going to have to work out how to have a hosted version elsewhere/turn it into a home-assistant "add-on" in docker.

**Reply by philosowaffle** [reply]
I assume by your comment that you are running HomeAssistant OS, which is a fantastic starting point, but it seems you're starting nudge up against the limitations of what it can do (i.e. having to jump through the hoops of "add-ons" just to run a new docker container).

You may find it beneficial to start poking at plain old Docker, it will open up a whole new world of possible software you can run without having to wait for HomeAssistant to have an add-on for it. 

But I do hear your pain points, I'm hoping a one-click deploy button to a free tier on some 3rd party service will be accessible for most people and fill the gap of GitHub Actions for those that need to have Two-Factor enabled.

Note for others who find themselves on this thread: GHA will continue to work so long as Garmin Two-Factor auth is disabled.

**Reply by qpop01** [reply]
I suspect I'll get mandatory 2FA soon as I have a fenix 7 which I believe has the hardware for ECG...

It looks like building a homeassistant add-on is just a case of specifying the details of a docker container : https://developers.home-assistant.io/docs/add-ons/tutorial/ 

If I got some help putting one together by anyone in the community, I'd happily upload it to HACS and make it publicly available for other HA users...obviously a free tier script on heroku would be an amazing alternative.

I asked a question on a fresh thread and then answered it (turning a manual sync into an iOS shortcut) which I think is a really neat thing to do, and hope it can be replicated with other methods.
